File: dotnet_distance_generator/trunk/src/csharp/main/TopCoder/Web/Distance/DistanceCalculators/GeographicalDistanceCalculator.cs
/*
 * Copyright (c) 2008, TopCoder, Inc. All rights reserved.
 */

using System;
using System.Collections.Generic;
using TopCoder.Web.Distance.Data;

namespace TopCoder.Web.Distance.DistanceCalculators
{

    /// <summary>
    /// This class is the implementation of the <see cref="IDistanceCalculator"/>
    /// interface that calculates the geographical distance. It implements the
    /// default geographical calculation formula as stated in the algorithm section
    /// of the component specification.
    /// </summary>
    /// <threadsafety>
    /// This class is thread-safe since it is immutable.
    /// </threadsafety>
    /// <author>TCSDESIGNER</author>
    /// <author>hotblue</author>
    /// <copyright>Copyright (c) 2008, TopCoder, Inc. All rights reserved.</copyright>
    /// <version>1.0</version>
    public class GeographicalDistanceCalculator : IDistanceCalculator
    {

        /// <summary>
        /// The default do-nothing constructor.
        /// </summary>
        public GeographicalDistanceCalculator()
        {
        }

        /// <summary>
        /// Calculates the geographical distances from the given member to all its
        /// related members. The returned list should contain distance to each
        /// related member in the same order as the related members are in the
        /// given list. A negative value in the returned list means that the
        /// distance is &quot;undefined&quot;.
        /// </summary>
        /// <param name="member">
        /// The member to calculate the distances for.
        /// </param>
        /// <param name="relatedMembers">
        /// All related members of the given member, cannot be empty list.
        /// </param>
        /// <param name="type">
        /// The competition types used to compute distance.
        /// </param>
        /// <returns>
        /// The geographical distances to each related member.
        /// </returns>
        /// <exception cref="ArgumentNullException">
        /// If either argument is null.
        /// </exception>
        /// <exception cref="ArgumentException">
        /// If the list is empty or contains a null element.
        /// </exception>
        public IList<float> CalculateDistance(Member member, IList<Member> relatedMembers, CompetitionTypes type)
        {
            Helper.ValidateNotNull("member", member);
            Helper.ValidateList<Member>("relatedMembers", relatedMembers, false);

            float[] result = new float[relatedMembers.Count];

            // Get the maximum distance of all members involved in the calculation including the origin member.
            float maxDistance = member.GeographicalDistance;

            foreach (Member relatedMember in relatedMembers)
            {
                if (relatedMember.GeographicalDistance > maxDistance)
                {
                    maxDistance = relatedMember.GeographicalDistance;
                }
            }

            if ((member.GeographicalDistance < 0f) || (maxDistance < 0f))
            {
                for (int i = 0; i < result.Length; i++)
                {
                    result[i] = -1;
                }
            }
            else
            {
                for (int i = 0; i < relatedMembers.Count; i++)
                {
                    if (relatedMembers[i].GeographicalDistance >= 0)
                    {
                        result[i] = (maxDistance > 0) ? (relatedMembers[i].GeographicalDistance / maxDistance) : 0f;
                    }
                    else
                    {
                        result[i] = -1;
                    }
                }
            }

            // Return the result
            return result;
        }
    }
}

File: java_distance_generator/trunk/src/java/main/com/topcoder/web/tc/distance/calculators/GeographicalDistanceCalculator.java
/*
 * Copyright (C) 2008 TopCoder Inc., All Rights Reserved.
 */
package com.topcoder.web.tc.distance.calculators;

import java.util.ArrayList;
import java.util.EnumSet;
import java.util.List;

import com.topcoder.util.errorhandling.ExceptionUtils;
import com.topcoder.web.tc.distance.Helper;
import com.topcoder.web.tc.distance.CompetitionType;
import com.topcoder.web.tc.distance.DistanceCalculator;
import com.topcoder.web.tc.distance.DistanceType;
import com.topcoder.web.tc.distance.data.Member;

/**
 * <p>
 * This implementation of the DistanceCalculator will calculate distance based on the geographical distance from
 * the coder.
 * </p>
 * <p>
 * This implementation will be created by the application and passed to the DistanceGenerator to calculate
 * distances based on geographical distance. The DistanceGenerator will then call this class with a list of members
 * to calculate. This class will then calculate the distance based on the geographical distance of those members
 * from the given coder.
 * </p>
 * <p>
 * Thread-safety: this class is thread safe by having no mutable state.
 * </p>
 *
 * @author Pops, romanoTC
 * @version 1.0
 */
public class GeographicalDistanceCalculator implements DistanceCalculator {
    /**
     * Default empty constructor.
     */
    public GeographicalDistanceCalculator() {
        // Does nothing
    }

    /**
     * This method returns the distance type associated with this calculator.
     *
     * @return always return {@link DistanceType#GEOGRAPHICAL}.
     */
    public DistanceType getDistanceType() {
        return DistanceType.GEOGRAPHICAL;
    }

    /**
     * <p>
     * This method will calculate, for each member, the geographical distance and return a list (of the same size
     * and order of members) for each member distance.
     * </p>
     *
     * @param compTypes
     *            A non-null enum set of competition types to use.
     * @param coder
     *            A non-null coder to get the distance from.
     * @param members
     *            A non-null, non-empty list of members to query.
     * @return A non-null list of the size members that contains the distance for that member the members order
     *         (the distance can be negative if undefined).
     * @throws IllegalArgumentException
     *             if any argument is null; if members is empty or contains null elements, if compTypes contains
     *             null elements; if members is null or an empty list.
     */
    public List<Double> getDistance(Member coder, List<Member> members, EnumSet<CompetitionType> compTypes) {
        ExceptionUtils.checkNull(coder, null, null, "The [coder] argument cannot be null.");
        Helper.checkCollection(members, "members", false);
        Helper.checkCollection(compTypes, "compTypes", true);

        List<Double> rcList = new ArrayList<Double>(members.size());

        int ourDistance = coder.getGeographicalDistance();
        int maxDistance = (ourDistance < 0) ? -1 : getMaxGeographicalDistance(members);

        for (Member member : members) {
            int memberDistance = member.getGeographicalDistance();

            // If members have no geographical info, add -1
            if (ourDistance < 0 || memberDistance < 0) {
                rcList.add(-1.0);

            } else {
                rcList.add((maxDistance == 0) ? 0.0 : ((double) memberDistance) / maxDistance);
            }
        }

        return rcList;
    }

    /**
     * <p>
     * Returns the max geographical distance from the list of members.
     * </p>
     *
     * @param members
     *            A non-null, non-empty list of members (with no null elements).
     * @return A double indicating the max distance (negative means unknown).
     */
    private int getMaxGeographicalDistance(List<Member> members) {

        int maxDistance = Integer.MIN_VALUE;

        for (Member member : members) {
            maxDistance = Math.max(maxDistance, member.getGeographicalDistance());
        }

        return maxDistance;
    }
}
